Fixes for deep and declared hierarchies

daumantas-kavolis-sensmetry opened the pull request:
Fixes #949:

- `$type` missing some subtype names occasionally - names are collected recursively
- assignment validation using inferred hierarchy which is very likely to be incomplete with declared types - replaced the inferred property type with a matching declared type for assignment checks only

This does not fix `$container` being non-optional on grammar entry types and its supertypes.

msujew commented on packages/langium/test/grammar/type-system/inferred-types.test.ts:
```diff
@@ -543,6 +543,44 @@ describe('types of `$container` and `$type` are correct', () => {
         `);
     });
+
+    test('types of `$type` for declared deep hierarchies', async () => {
```
Suggestion: It might make sense to split this test into multiple tests that test for the following scenarios:

1. Linear (`C extends B, B extends A`)
2. Deviating (`B extends A, C extends A`)
3. Joining (`C extends A and B`)
4. A mix of all of the above.

msujew commented on packages/langium/test/grammar/type-system/type-validator.test.ts:
```diff
@@ -497,4 +497,27 @@ describe('Property types validation takes in account types hierarchy', () => {
 
+    test('No false positive on declared type assignment', async () => {
+        const validation = await validate(`
+        interface A {}
+        interface B extends A {}
+        interface C extends B {}
+        interface D extends C {}
+        interface E extends C {}
+        interface F extends D {}
+        interface G extends D {}
+        interface H extends A {}
```
Suggestion: I feel like there are a few interfaces in here which only add noise. We can probably just trim  of them down to `A, B, C` and accomplish the same effect.

msujew reviewed (commented):
Thanks, that's very helpful! I have a few improvements on the testing code, but otherwise this looks already pretty good 👍

daumantas-kavolis-sensmetry commented:
I have split the `$type` tests into smaller ones but only the last one failed without the fix, and removed as much as possible from the assignment test such that it failed without a fix.

daumantas-kavolis-sensmetry commented:
Wait, the `$type` test doesn't fail on `main`, I'll look into it.

msujew reviewed (approved):
Looks good to me 👍
